Update feature/coupled-crow to run the latest ufs-weather-model with prognostic aerosols

rmontuoro opened the pull request:
This PR replaces the original coupled atmosphere-aerosols system (FV3-GOCART) with the latest revision of the ufs-weather-model including prognostic aerosols. The model's code is temporarily retrieved from a forked repository as the corresponding updates are being integrated in the authoritative ufs-weather-model repository. Aerosol test cases are also updated, and the original `cplchem` flag is reused to enable coupling with the aerosol component, removing the additional `cplgocart` flag.

The included changes partially address #324, updating FV3 namelist settings to run the latest FV3 revision with the `FV3_GFS_v16` CCPP suite used with the coupled UFS Aerosols component.
Commit d16630f3b34671e2350e6a88b9df749e4fd58ece also fixes a bug reported in #337.

Closes: #336, #337
Refs: #324

rmontuoro committed:
Checkout and build ufs-weather-model app integrating
prognostic aerosols into the UFS, replacing original
FV3-GOCART system.

A forked repository is temporarily used to download
the updated ufs-weather-model app.

rmontuoro committed:
Use proper modulefiles for updated ufs-weather-model
app with prognostic aerosols.

rmontuoro committed:
Update parsing function for FV3 namelist to work
with latest FV3 revision and include settings for
FV3_GFS_v16 and aerosol scavenging factors.

rmontuoro committed:
Add aerosol scavenging factors (fscav_aero) to FV3 settings.
This item is implemented as a YAML list of strings, which
allows to set scavenging factors for each tracer.

rmontuoro committed:
Properly set NST_SPINUP flag in FV3 namelist.

rmontuoro committed:
Restore `cplchem` as the flag for coupled aerosols,
removing all references to `cplgocart`.

rmontuoro committed:
- Set NEMS identifier for coupled aerosol component to 'gocart'.

- Copy NUOPC field dictionary to run directory by default.

rmontuoro committed:
Update test cases and input files for coupled aerosols runs.

rmontuoro committed:
Add scavenging factors to aerosol test cases.

WalterKolczynski-NOAA commented:
While I'm not necessarily against the changes, there are important changes here (like changing the modulefile used by global-workflow at runtime) that really should've been discussed with the global-workflow team before being implemented. Additionally, the module name change impacts the non-aerosol version, which is still using the old UFS naming convention. Finally, with regard to the modulefile, I don't know that UFS is going to want to keep a separate modulefile for aerosols, given that no other component has its own modulefile in the root modulefile directory.

The change from cplgocart to cplchem in CROW is also going to conflict with PR #334, which makes changes keyed off of cplgocart that will need to be updated.

On the whole, I'd prefer to sit on this one if possible at least until the model changes are merged into the authoritative ufs-weather-model repo and we can coordinate with @JessicaMeixner-NOAA to update the entire coupled workflow to the latest UFS commit at the same time. Ideally merging to a single build command with all the components compiled as well, instead of different build settings.

JessicaMeixner-NOAA commented:
Since this only partially addresses #324 I'm afraid if we merge this by itself we will break the ATM-OCN-ICE-WAV model workflow.  However, if we only address #324 without this, we'd likely break the aerosols app as well.  So perhaps these will have to be combined into an effort.  I'm still not sure who is going to address #324 in full but hopefully we can discuss that at the tag-up here shortly.  

@WalterKolczynski-NOAA Depending on how long it will take to fully reintegrate GOCART into ufs-weather-model, I don't think waiting until it's one build system should stop us.  However, if we're changing modules for the ATM-OCN-ICE-WAV as well, we really need to consider how that impacts. 

@rmontuoro did you run the coupled_free_forecast_wave.yaml case? Perhaps I'm thinking it conflicts with #324 when it actually doesn't?

JessicaMeixner-NOAA commented on sorc/build_ufs_coupled.sh:
```diff
@@ -13,15 +13,16 @@ if [ $target = hera ]; then target=hera.intel ; fi
+module load ufs_${target}
 cd ufs_coupled.fd/
 if [[ -d build ]]; then rm -Rf build; fi
 if [[ -d GOCART ]]; then
```
So you have switched to using your fork, but as soon as this PR goes into ufs-weather-model this build check will not work anymore.  Do we address that now or later?

rmontuoro replied:
In the most recent versions of `ufs-weather-model`all modulefiles have been renamed as: `ufs_<machine>.<compiler_settings>`. This change introduces the update and should work from now on.

JessicaMeixner-NOAA replied:
But you didn't update the hash of what is checked out for the ufs-weather-model for the coupled case.  Therefore this will break the coupled case as is,  unless I missed that.

WalterKolczynski-NOAA replied:
This was my issue as well. The currently non-aerosol coupled-crow ufs model version is still using the old naming convention. So this PR would break that until #324 is also merged.

WalterKolczynski-NOAA replied:
(And vice versa)

JessicaMeixner-NOAA replied:
Well if the changes were only within the if/else statements it could be fine, but as it is this will break atm-ocn-atm-wav

rmontuoro replied:
You are right. My assumption was that we were going to update to the latest ufs-weather-model for the coupled case, too, before or at the same time as this PR. I will address this issue by adding a conditional block.

JessicaMeixner-NOAA replied:
I hope we will update ufs-weather-model for the other case very soon, but as of now #324 is remains an unassigned task.

rmontuoro replied:
Commit f81741bdb3a8cc221f6e0846c706cfe3cc9c4010 addresses this issue. Updated modulefiles are only loaded to build the ufs-weather-model with prognostic aerosols. This model is identified though a hidden file, created when using option `-a` during checkout.

WalterKolczynski-NOAA replied:
Looks like I'm too late and the complications have already been added, but I'm going to try and tackle #324 after we clear a couple of the other PRs out. So let's cool it with adding more if-block exceptions for aerosols when we want to be removing them.

WalterKolczynski-NOAA replied:
At least until we can discuss it in the afternoon.

JessicaMeixner-NOAA commented on ush/forecast_postdet.sh:
```diff
@@ -868,7 +864,7 @@ GSD_in() {
 GSD_nml() {
 	echo "SUB ${FUNCNAME[0]}: Creating name list for GSD"
-	sh parsing_namelists_GSD.sh
+	# generate GSD-specific input
```
should this whole section just be deleted now?

rmontuoro replied:
Agreed. I was not sure if we wanted to leave it as a template for future development.

rmontuoro replied:
Commit 6b19a39b6f086ff7eb077c60a3acbc40d6dae5a1 removes such GSD-specific shell functions.

JessicaMeixner-NOAA commented on ush/load_fv3gfs_modules.sh:
```diff
@@ -24,10 +24,10 @@ elif [[ -d /lfs3 ]] ; then
 elif [[ -d /scratch1 ]] ; then
 	# We are on NOAA Hera
-	module load module_base.hera
+	module use "$HOMEgfs/sorc/ufs_coupled.fd/modulefiles"
```
I don't necessarily disagree with this change, but @WalterKolczynski-NOAA needs to be okay with it.  I don't think there are other modules that are needed when running the forecast but I could be wrong.  This is just for the forecast job right? If it's more than just the forecast I don't think this will work.

rmontuoro replied:
I fully defer to your choice here. My assumption was that platform-specific modulefiles provided with the source code would include all necessary modules to run different coupled configuration but this may not be the case.

While this is mostly a convenience-driven change, I'm concerned about duplicating model settings in the workflow as it will make it harder to maintain.

WalterKolczynski-NOAA replied:
The global-workflow team has already discussed moving to this paradigm (using the same script for build and runtime), but haven't gotten to it yet. Here I want to be sure that replacing base with the UFS module here is only impacting the forecast job and not any other jobs.

rmontuoro replied:
Commit 0307896af1d309718d6b08d89631f8c080c575e7 refactors this script to:
- use a new shell function (`get_platform`) introduced to return the string identifier for the auto-detected local platform to improve modularity
- load updated modulefiles only for coupled aerosols on Hera/Intel

rmontuoro replied:
Commit a821a1aef761b386beace4efab476495e209adbb fixes additional issues with the script.

JessicaMeixner-NOAA commented on ush/load_fv3gfs_modules.sh:
```diff
@@ -24,10 +24,10 @@ elif [[ -d /lfs3 ]] ; then
 	if [[ -d $HOMEgfs/sorc/ufs_coupled.fd/GOCART ]] ; then
-		module use "$HOMEgfs/sorc/ufs_coupled.fd/modulefiles/hera.intel"
-		module load fv3
+		module load ufs_aerosols_hera.intel
```
Like above, this works for now, but this will not be sufficient after the PR goes in.

rmontuoro replied:
This modulefile should go away soon.

JessicaMeixner-NOAA replied:
The issue is the if the GOCART directory exists is not going to work once the ufs-weather-model PR is committed and all scenarios have the folder GOCART.  So this will have to not adversely affect the other scenarios.

rmontuoro replied:
Agreed. We should limit use of the additional aerosol modulefile only to coupled configurations building GOCART.

rmontuoro replied:
See comment above.

rmontuoro replied:
Work is ongoing to include all GOCART dependencies into hpc-stack.

JessicaMeixner-NOAA commented on ush/nems_configure.sh:
```diff
@@ -155,10 +155,10 @@ EOF
-
 fi 
 
+cp $HOMEgfs/sorc/ufs_coupled.fd/tests/parm/fd_nems.yaml fd_nems.yaml
```
This is rightfully not just under cplflx, but should it be behind a cpl flag? I guess it doesn't hurt standalone atm, its just not really needed?

rmontuoro replied:
This is an interesting question.

Yes, the NUOPC field dictionary is not required by standalone atm. Furthermore, recent FV3 versions no longer rely on the `cpl` flag, formerly in the `model_configure` file, to advertise/realize coupling fields in the NUOPC cap. We could wrap this file copy in a conditional block to check whether any of the physics coupling flags (`cplflx`, `cplwav`, etc.) are true, similarly to what implemented in `cplvalidate.sh`:
https://github.com/NOAA-EMC/global-workflow/blob/3428097dd39934ea648d86e5cdee8d8c2bdb3b14/ush/cplvalidate.sh#L32
However, there is an ongoing effort to refactor the FV3 cap, with the possible (but remote) outcome that other coupling flags may no longer be used, too.

JessicaMeixner-NOAA replied:
This isn't just used by the atm model in coupling. It's used by any model.  Did the "cpl" flag get removed completely? I believe it's still a variable in the workflow if not in input.nml/model_configure 

At any rate,  it doesn't adversely affect fv3 standalone so it can just be left as is for simplicity.

rmontuoro replied:
Yes, the `cpl` flag was removed and it is no longer needed in `model_configure`.

JessicaMeixner-NOAA commented on workflow/cases/aerosol_firex_forecast.yaml:
```diff
@@ -5,8 +5,8 @@ case:
     BASE_CPLIC: /scratch1/NCEPDEV/nems/Raffaele.Montuoro/data
-    OROFIX:     /scratch1/NCEPDEV/nems/emc.nemspara/RT/NEMSfv3gfs/develop-20200914/INTEL/FV3_input_data_c384/INPUT
-    GRDFIX:     /scratch1/NCEPDEV/nems/emc.nemspara/RT/NEMSfv3gfs/develop-20200914/INTEL/FV3_input_data_c384/INPUT
+    OROFIX:     /scratch1/NCEPDEV/nems/emc.nemspara/RT/NEMSfv3gfs/input-data-20210528/FV3_input_data384/INPUT
```
Should these be put with the global-workflow fix files instead? @KateFriedman-NOAA would be the person to coordinate that update with.

rmontuoro replied:
I'll follow your guidance on this, too. Note the updated baseline linked to the updated ufs-weather-model.

JessicaMeixner-NOAA replied:
The workflow does not use the rt.sh inputs and instead use the global-workflow set. This is fine for now, but we need to get whatever is needed here in the global-workflow fix files that @KateFriedman-NOAA is maintaining.

WalterKolczynski-NOAA replied:
Yes, all fix files should be in the fix directory. Nothing should look into the UFS regression test directory as we clean things up. I'm okay addressing this later though.

rmontuoro replied:
References to aerosol-specific fix files have been removed.

JessicaMeixner-NOAA reviewed (commented):
Made a few comments.  Before this is merged, we need to confirm that the atm-ocn-ice-wav case still runs as is.

junwang-noaa commented:
@*rmontuoro *All the UFS applications require this yaml file including the
standalone FV3 since May 21 So it's good to keep the yaml file without any
constraints.

On Wed, Jun 9, 2021 at 4:02 PM Raffaele Montuoro ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In ush/nems_configure.sh
> <https://github.com/NOAA-EMC/global-workflow/pull/338#discussion_r648644365>
> :
>
> >  fi
>
> +cp $HOMEgfs/sorc/ufs_coupled.fd/tests/parm/fd_nems.yaml fd_nems.yaml
>
> This is an interesting question.
>
> Yes, the NUOPC field dictionary is not required by standalone atm.
> Furthermore, recent FV3 versions no longer rely on the cpl flag, formerly
> in the model_configure file, to advertise/realize coupling fields in the
> NUOPC cap. We could wrap this file copy in a conditional block to check
> whether any of the physics coupling flags (cplflx, cplwav, etc.) are
> true, similarly to what implemented in cplvalidate.sh:
>
> https://github.com/NOAA-EMC/global-workflow/blob/3428097dd39934ea648d86e5cdee8d8c2bdb3b14/ush/cplvalidate.sh#L32
> However, there is an ongoing effort to refactor the FV3 cap, with the
> possible (but remote) outcome that other coupling flags may no longer be
> used, too.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/global-workflow/pull/338#discussion_r648644365>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AI7D6TNJT6PCYBGW7UTT46DTR7CERANCNFSM46LHFZVQ>
> .
>

rmontuoro committed:
Use updated modulefile names only when building the
ufs-weather-model with prognostic aerosols. This
option is recognized by means of a hidden file
created in the model's repository when running
checkout.sh with option -a.

rmontuoro committed:
Load updated modulefiles for coupled aerosols on Hera/Intel.

A new shell function (get_platform) is introduced to increase
modularity. This fuction auto-detects the local platform and
returns a string identifier. As a result, load_fv3gfs_modules.sh
is refactored.

rmontuoro committed:
Remove GSD-specific functions in forecast postdet.

rmontuoro committed:
Remove no longer needed parsing script used to
generate the FV3 namelist file for FV3-GOCART.

rmontuoro committed:
Return proper name for Hera platform and load correct modules.

rmontuoro commented:
@JessicaMeixner-NOAA - the `coupled_free_forecast_wave.yaml` case runs successfully after applying the changes included in this PR.

rmontuoro committed:
Load aerosol-specific modulefiles only for the forecast task.

rmontuoro committed:
Merge branch 'WalterKolczynski-NOAA:feature/aerosol_restart' into feature/update-aerosols

rmontuoro committed:
Replace legacy 'cplgocart' flag with 'cplchem'.

rmontuoro committed:
Update tracer list.

rmontuoro committed:
Merge branch 'WalterKolczynski-NOAA:feature/aerosol_restart' into feature/update-aerosols

rmontuoro committed:
Restore original path to modulefiles in module_base.hera.

rmontuoro committed:
Merge 'WalterKolczynski-NOAA:feature/aerosol_restart' into feature/update-aerosols

rmontuoro committed:
Update aerosol_firex_forecast test case.

rmontuoro commented:
This branch now includes the latest updates from PR #334

WalterKolczynski-NOAA committed:
Update UFS model hash

Updates the UFS model to the current tip of develop.

Refs: #324

WalterKolczynski-NOAA committed:
Update MOM_input with landmask change option

Adds the ALLOW_LANDMASK_CHANGES option to the MOM6 input templates
and allows the setting via MOM6_ALLOW_LANDMASK_CHANGES (default
False).

Refs: #324

WalterKolczynski-NOAA committed:
Update CICE inputs for new model version

The newest version of CICE replaces istep0 with the normal month and day,
so the namelist is updated to reflect that.

ktherm and tfrz_option are updated to allow alternate settings, following
similar changes to the UFS test harness. The default values are the
previous settings.

Settings for tr_FY, sw_redist, f_mlt_onset, and f_frx_onset are updated
to new values in the UFS test harness.

Additionally, many namelist options have been removed following the same
deletions in the UFS test harness.

Refs: #324

WalterKolczynski-NOAA committed:
Update UFS build script

Some changes to the UFS require updates to the build script.

The modulefile organization has been updated to be
modulefiles/ufs_${target} rather than modulefiles/${target}/fv3, so
the module paths in the build script are updated correspondingly.

A cmake switch DAPP is now required in order to build UFS. Non-
aerosol coupled is using S2SW for the time being. This replaces
setting DS2S and DWW3 to YES.

The old v15 suite is removed from the CCPP suite list.

Refs: #324

WalterKolczynski-NOAA committed:
Update source of fd_nems.yaml

The location of the fd_nems.yaml file changed in the UFS repository,
so updated nems_configure to copy from the new location.

Refs: #324

WalterKolczynski-NOAA committed:
Update templates to set ocean mesh file

The latest version of UFS adds a setting to define the ocean mesh file,
so the nems.configure templates are updated to define them. They are
identical to the ice mesh file, so the MESHICE variable is renamed to
MESH_OCN_ICE and used for both. Additionally, the use_mommesh variable
is added and defaults to true.

Refs: #324

WalterKolczynski-NOAA committed:
Update Ww3 prepost build

The WW3 pre and post components now need NetCDF to build, so NetCDF is
added back to the modulelist, and previously removed code in the build
script is restored.

Refs: #324

JessicaMeixner-NOAA committed:
update wave ww3_multi.inp file Issue #325

WalterKolczynski-NOAA committed:
Update meditaor name

In the latest UFS, the mediator name was changed from nems to cmeps, so
the nems.configure is updated to match.

Refs: #324

WalterKolczynski-NOAA committed:
Update wave IC

Update wave IC to new set compatible with current UFS version.

Refs: #324

WalterKolczynski-NOAA committed:
Update modules to match UFS

Update module versions to match those in the UFS common file.

Refs: #324

WalterKolczynski-NOAA committed:
Update ice configuration to match UFS

There were still a number of settings in the ice input file that
did not match the setting in UFS (and one setting that was misplaced).

Refs: #324

WalterKolczynski-NOAA committed:
Merge branch 'feature/coupled-crow' into feature/model_update

rmontuoro committed:
Merge 'feature/coupled-crow' into feature/update-aerosols

rmontuoro committed:
Add ammonia and nitrates to the list of cycled GOCART tracers.

rmontuoro committed:
Update checkout and build script to use the authoritative
GitHub repository for the UFS weather model only. This repo
now contains UFS-Aerosols, the aerosol component based on
NASA's GOCART.

Since we are using a single repository for both S2S and
coupled aerosols, the -a option was removed from checkout.sh
and added to build_all.sh. This option is passed to
build_ufs_coupled.sh and partial_build.sh, which have
been streamlined and now accept both short and long options:

build_ufs_coupled.sh:
  -a, --aerosols
  -c, --coupled

partial_build.sh:
  -a, --aerosols
  -c, --coupled
  -h, --help
  -v, --verbose

KateFriedman-NOAA committed:
Add spaces to indent FV3 build section for better readability

Refs: #230

KateFriedman-NOAA committed:
Add FV3_GFS_v16_couplednsst CCPP suite option to UFS coupled build and FV3 namelist parser

Refs: #230

KateFriedman-NOAA committed:
Replace CROW yaml nst_spinup_logic in nsst.yaml with NST_SPINUP variable

- removed CROW FirstTrue check when building config.nsst from nsst.yaml
- replace nst_spinup_logic with NST_SPINUP variable that is parsed from variable dictionary
- nst_spinup_logic not needed currently

Refs: #230

KateFriedman-NOAA committed:
Add NSST settings to the coupled_free_forecast_wave.yaml case file

- set NST_MODEL=2 (turns it on)
- add additional NSST namelist variables: NST_SPINUP, NST_RESV, ZSEA1, ZSEA2
- set NST_SPINUP=1 (NSST will spinup)
- change CCPP_SUITE from FV3_GFS_v16_coupled to FV3_GFS_v16_couplednsst

Refs: #230

WalterKolczynski-NOAA committed:
Update nems.configure to inclue eps_imesh

Adds the eps_imesh setting to the nems.configure files when CICE is
on.

Refs: #324, #333

WalterKolczynski-NOAA committed:
Merge branch 'feature/model_update' into feature/update-aerosols

WalterKolczynski-NOAA committed:
Change forecast module back to module_base

The modulefile in UFS doesn't load some modules needed for the forecast
script at runtime, notably produtil, which provides err_chk. The module
for the forecast job is changed back to module_base for the time being.

Refs: #336

WalterKolczynski-NOAA commented:
This PR is being rolled into PR #359 after all, as I needed the changes to the checkout/build for the updated UFS.
